server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java

```diff
@@ -317,38 +319,33 @@ public void decrypt(char[] password) throws GeneralSecurityException, IOExceptio
              DataInputStream input = new DataInputStream(bytesStream)) {
```

Comment:
I think we should add a test that would have failed using the old method. This would probably require a small refactoring to do this; I am thinking instead of `new ByteArrayInputStream` we call a method that could be overridden in a test. There we would have a special stream where a call to `read` would not fully populate the byte array

Reply:
I don't think we need to override any methods. We can form streams of bytes in whatever order we want in tests.
However, if I understand the problem correctly, it has to do with read boundaries. Eg, an input stream may choose to return less bytes to a `read` call than the array can contain, since it may just copy the rest of an existing buffer (eg a decrypted block in CipherInputStream). By calling `readFully`, multiple read calls are made to ensure the array is filled completely or EOF is reached. I think this will be very difficult to mimic, given the behavior of `read` here will be implementation dependent. The best we could do is add tests to trigger each of the possible EOFs from `readFully` calls by truncating the keystore at certain points (although this would probably be difficult to do within the encrypted bytes).

I do think we should have a test to ensure there cannot be garbage after the encrypted data.

Reply:
Revisiting this after coming back. 
> I do think we should have a test to ensure there cannot be garbage after the encrypted data

Care to elaborate @rjernst ? You mean attempting to `readFully()` into a larger array and verifying that we get the EOFException rather than keep reading garbage ?

Reply:
I mean adding a test with garbage at the end, so that we know the last readFully logic works. The intermediate readFully's are harder, as I described above.

Reply:
> I do think we should have a test to ensure there cannot be garbage after the encrypted data.

If I understand correctly, we should fail if there is garbage right? Currently this is not the case at least based on the test @jkakavas added.

Reply:
`readFully` reads exactly the number of bytes we instruct it to, or throws an EOFException. I interpreted @rjernst comment above to mean that we need to have a test verifying that even if there  more garbage, readFully won't read them, but read the bytes it should(this is what the test I added does, and it should not fail ) 
Another way to interpret the comment I guess, would be to write less bytes to would be to write fewer bytes to the `DataOutputStream` than we indicate with the `writeInt` above and make sure that on decryption, readFully throws the EOFException instead of reading in garbage in order to fill our byte array - I can add that too.
Writing about this, the second interpretation makes much more sense I guess..

Reply:
That was not my intention. I meant it the way @jaymode suggested: we should fail with garbage, and have a test for that.